Amittal94/unified flatbuffer

amittal94 opened the pull request:
This branch adds the following:
A pva0 fbs schema based on normative types and pvData.
Flatc compiled code for this schema.
Python serialisers and deserialisers for this schema. It relies on following defined pydantic classes.
Partial support for channel access based scalars under the assumption that they are sent in the corresponding format.

amittal94 committed:
define pva0 fbs and add generated files

amittal94 committed:
define pydantic types and pva0 serialiser and deserialisers

amittal94 committed:
support for ca scalar type and conversion

amittal94 commented:
Some questions:
How do I handle typing in the serialise_any and deserialise_any portion?
As discussed the conversion of CA data may be better handled here, but I am not convinced of my approach to this currently? Any CA scalar data can be sent as PVData object with data of CAScalarAll type and it will be converted and serialised as a NTScalarAll type.

stephen-jd-dann commented:
I still need to go through this properly, but...

> Some questions: How do I handle typing in the serialise_any and deserialise_any portion?

I guess the question is how to represent the `any` type in Python? For `serialise_any` I think we just use `numpy.typing.ArrayLike`, which seems to include scalars. We might need to check whether strings work. We might be able to use the same type for `deseiralise_any`, but that's a bit more specifically `int | float | str | np.ndarray`.

> As discussed the conversion of CA data may be better handled here, but I am not convinced of my approach to this currently? Any CA scalar data can be sent as PVData object with data of CAScalarAll type and it will be converted and serialised as a NTScalarAll type.

If I understand this correctly, the steps are:

1. Read data with `pyepics`
2. Convert to `CAScalarAll`
3. Convert to `NTScalarAll` with `cascalarall_to_ntscalarall`
4. Do stuff with it e.g. serialise it

We'd need to check this works nicely with both `epics.PV.get()` and the callbacks used for monitor. Even then, the extra type might not help client code much, so having a method that takes a dictionary might be simpler.

amittal94 commented:
> We'd need to check this works nicely with both `epics.PV.get()` and the callbacks used for monitor. Even then, the extra type might not help client code much, so having a method that takes a dictionary might be simpler.

It will require some handling.  On the forwarder (client code?) the way I have been testing is that I capture the kwargs passed to the monitor callback and then initialise the CAScalarAll with the kwargs. 'get_with_metadata() ', the closest analogue to the callback, can be handled the same way. The flow is similar with p4p, using a pvdata.todict() and sending it as kwargs to NTScalarAll directly. Presuming you are referring to CAScalarAll as the extra type, in the former case we can directly use the dictionary and adapt the conversion function accordingly. Part of the reason I made a type was under the consideration that it gives a standard to follow, if say we used something else other than pyepics in the future, and want to make it compatible as well to reduce human error via mapping to attributes.

stephen-jd-dann commented:
> > We'd need to check this works nicely with both `epics.PV.get()` and the callbacks used for monitor. Even then, the extra type might not help client code much, so having a method that takes a dictionary might be simpler.
> 
> It will require some handling. On the forwarder (client code?) the way I have been testing is that I capture the kwargs passed to the monitor callback and then initialise the CAScalarAll with the kwargs. 'get_with_metadata() ', the closest analogue to the callback, can be handled the same way. The flow is similar with p4p, using a pvdata.todict() and sending it as kwargs to NTScalarAll directly. Presuming you are referring to CAScalarAll as the extra type, in the former case we can directly use the dictionary and adapt the conversion function accordingly. Part of the reason I made a type was under the consideration that it gives a standard to follow, if say we used something else other than pyepics in the future, and want to make it compatible as well to reduce human error via mapping to attributes.

That's a good point. We may want to add more specialised conversion methods in the future for `pyepics`, `ophyd`, and others.

In any case, having a data type to represent the CA data is good, although it probably shouldn't be the main/only API the user sees.

stephen-jd-dann commented on pyproject.toml:
```diff
@@ -7,6 +7,7 @@ name = "epac-flatbuffer-formats"
 dependencies = [
     "flatbuffers==23.5.26",
     "numpy~=2.0",
+    "pydantic==2.10.6"
```
No need to pin this (unlike flatbuffers).

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,138 @@
+    BoolArray, ByteArray, UByteArray, ShortArray, UShortArray, IntArray, UIntArray, LongArray, ULongArray, FloatArray, DoubleArray, StringArray
+}
+
+table AnyOuter {
```
I see what you're doing, but
```suggestion
table Any {
```
will read better

stephen-jd-dann replied:
Re-reading this, I see that you'll get a naming conflict with `typing.Any`. `AnyOuter` still looks weird.

amittal94 replied:
It is indeed unusual, just the first thing I came up with when I ran into the conflict. We can rename if you wish, maybe AnyT to match others used throughout.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,138 @@
+
+table EnumT {
+  index: int;
+  choices: [string];
```
It might be better to only record the selected choice? Not a direct mapping of normative types, but it's less stupid IMO.

stephen-jd-dann replied:
As you'll see below, I now think that `EnumT` isn't pulling its weight and we should just get rid of it.

amittal94 replied:
We can do so, will need some clientside logic, since the PVData from p4p follows this convention. May need to discuss this further.

amittal94 replied:
I did come up with an approach using a custom validator, but the question would be to always use that validator or not.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,138 @@
+  display: DisplayT;
+}
+
+table NTScalarAll {
```
Still not entirely happy with this name, but I don't have a better idea.

amittal94 replied:
ScalarT or just Scalar perhaps? Accordingly could change the other names too.

amittal94 replied:
switched to NTScalarAny as discussed

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -0,0 +1,30 @@
+    alarm = dt.AlarmT(severity=data.severity, status=data.status)
+
+    timeStamp = dt.TimeT(
+        secondsPastEpoch=int(data.timestamp),
+        nanoseconds=int((data.timestamp - int(data.timestamp)) * 1e9),
```
Oh that's nasty. CA gives us the timestamp in this format already, `pyepics` convert it to a float for us, and now we have to convert it back.

Perhaps `TimeT` should just contain a floating point number of seconds?

amittal94 replied:
Yeah I dont like this either, but there is a discrepancy between pyepics and p4p on returned value here.

stephen-jd-dann replied:
I think we leave this as it is. It's ugly but not really avoidable.

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -0,0 +1,30 @@
+        limitLow=data.lower_disp_limit,
+        limitHigh=data.upper_disp_limit,
+        units=data.units,
+        precision=data.precision or dt.DisplayT().precision,
```
What if `data.precision` is 0? Is the default 0? If so, you might as well write `data.precision or 0`.

Aside from that, not a big fan of the way this is done, but the only other option is messing around with keyword argument dictionaries.

amittal94 replied:
Way I see it, defaults should be only defined in one place, so we dont end up with a situation where we have 0 here, and some other value in the DisplayT class for instance. We can explore other options, will let you know if I think of anything.

stephen-jd-dann replied:
The point is that this code is only correct if the default is 0. If the default is 2 and `data.precision` is 0, the `or` will evaluate to the second option, 2, when it should be 0. You really need the proposed None-coalescing operator.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,116 @@
+    message: str = ""
+
+
+class TimeT(BaseModel):
```
You can add methods to these (I assume). A good pair here would be to get the timestamp as a float and as a UTC `datetime`.

amittal94 replied:
If our goal is to actually avoid deserialising later, it may be moot, but yes should be possible to add these methods.

stephen-jd-dann replied:
Yeah, no point adding them until they're needed.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,116 @@
+    size: int = 0
+    offset: int = 0
+    fullSize: int = 0
+    binning: int = 1
```
While this ought to be the default, it isn't in fbs and I don't think it is in PVData either. This should probably match???

If the intention is for the user to construct these directly, we should probably remove the defaults from `size` and `fullSize`.

amittal94 replied:
added default to the flatbuffers schema for binning

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,116 @@
+    description: str = ""
+    units: str = ""
+    precision: int = 0
+    form: Optional[EnumT] = None
```
This is a specific enum; we might want to define it as one rather than using `EnumT`.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
`EnumT` makes me sad. The PVData definition of it is a bit wasteful. It only seems to be used in one place. Perhaps we can get rid of it? From the schema as well.

amittal94 replied:
I will see what I can do

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+from typing import Callable, Optional, TypeVar
+from epac.flatbuffers.ca_to_pva import cascalarall_to_ntscalarall
+import flatbuffers
+import epac.flatbuffers.data_types as dt
+from functools import wraps
```
If you're not careful I'll turn `isort` on.

amittal94 replied:
I will improve it based on pep8

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+)
+import numpy as np
+
+map_dtype_to_scalar_fb = {
```
The regular naming here suggests you may be able to build these dynamically with fewer lines of code.

amittal94 replied:
partially improved

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+)
+import numpy as np
+
+map_dtype_to_scalar_fb = {
```
Consider hiding the dictionaries and requiring access through functions. That way if we have to make more complex decisions based on the dtype we can.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    return wrapper
+
+
+def _serialise_string_scalar(builder: flatbuffers.Builder, data: np.ndarray):
```
Return types would be good, but not essential.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+            data_start_offset, fb_type, enum_type = _serialise_numeric_scalar(
+                builder, data
+            )
+    else:  # Array
```
I'd limit this to `ndim == 1` and throw an error for other cases. The shape information will be lost, and the user can always call `.flatten()` if that's what they want.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+    # Handle empty arrays
+    if data.ndim > 0 and data.size == 0:
+        raise ValueError("Cannot serialise empty arrays.")
```
Why not?

amittal94 replied:
not sure why I did that, changed to handle it.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    Raises:
+        ValueError: If an attempt is made to serialise an empty array.
+    """
+    data = np.array(data)
```
```suggestion
    data = np.asarray(data)
```

I think this avoids a copy if `data` is already an array. I checked and it works when passed objects too.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+    if data.ndim == 0:  # Scalar
+        if np.issubdtype(data.dtype, np.str_) or np.issubdtype(data.dtype, np.bytes_):
+            data_start_offset, fb_type, enum_type = _serialise_string_scalar(
```
You could shorten this code even more by assigning the function to a variable and calling it at the end. Is that too clever?

amittal94 replied:
which function?

stephen-jd-dann replied:
The idea was that this would be `serialise_func = _serialise_string_scalar`, then after the ifs you'd have `data_start_offset, fb_type, enum_type = func(...)`. But that's unnecessarily complicated.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+
+def _serialise_string_scalar(builder: flatbuffers.Builder, data: np.ndarray):
+    """serialises a single string scalar into FlatBuffers format.
```
I like doc strings, but these may be a bit much for functions that are used only once.

Perhaps a comment above the first one explaining the common signature and the fact they're used below in `serialise_any`?

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    AnyOuter.Start(builder)
+    AnyOuter.AddValueType(builder, enum_type)
+    AnyOuter.AddValue(builder, data_offset)
+    return AnyOuter.End(builder)
```
Look how much overhead we need for a simple variant type!

Flatbuffers is really annoying sometimes.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    return AnyOuter.End(builder)
+
+
+def deserialise_any(buffer):
```
```suggestion
def deserialise_any(buffer: AnyOuter):
```
is this right?

amittal94 replied:
It does not work out of the box, but I will investigate more.

stephen-jd-dann replied:
I've given this a go myself. It's now `AnyT.AnyT`. The problem seems to be that `map_any_scalar_enum_to_type` is inferred as `dict[int, object]`. You can annotate it as `dict[int, Any]`. The next problem is that `data_offset` might be `None`, but checking for that and returning `None` seems to work.

Frankly, I think the only way to be reasonably sure that we've dealt with all possible malformed or weirdly formed flatbuffers data is with a fuzzer, and that's probably not worth the effort.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+    if data_enum in map_any_scalar_enum_to_type:
+        data_fb = map_any_scalar_enum_to_type[data_enum]()
+        data_offset = buffer.Value()
+        data_fb.Init(data_offset.Bytes, data_offset.Pos)
+        data = data_fb.Value()
```
Seriously!? This is what you have to do!?

amittal94 replied:
AFAIK.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+        if data_enum == AnyInner.AnyInner.StringArray:
+            data = np.array(
+                [str(data_fb.Value(n), "utf-8") for n in range(data_fb.ValueLength())]
+            )
```
Ooh, one of my favourite tricks:

```suggestion
            data = np.fromiter(
                str(data_fb.Value(n), "utf-8") for n in range(data_fb.ValueLength()),
                str,  # This needs to be the numpy string dtype and I can't remember how it works
                data_fb.ValueLength(),  # Not necessary but I think this slightly helps performance
            )
```

amittal94 replied:
I would need to specify a max length apparently.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+
+@safe_serialise
+def serialise_alarm(builder: flatbuffers.Builder, alarm_data: dt.AlarmT) -> int:
```
I hope you didn't have to write these all out by hand...

amittal94 replied:
not completely...

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+        EnumT.AddChoices(builder, choices_vector_offset)
+        form_offset = EnumT.EnumTEnd(builder)
+    else:
+        form_offset = 0
```
Nasty. I think we definitely need to try and get rid of `EnumT`.

amittal94 replied:
I took an approach relying on the pydantic types that does work, but we should assess it further.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+        form: dt.EnumT = display_data.form
+        choices_offsets = [builder.CreateString(choice) for choice in form.choices]
+        EnumT.StartChoicesVector(builder, len(choices_offsets))
+        for choice_offset in reversed(choices_offsets):
```
You have to do them backwards!?

amittal94 replied:
actually , I need to test this, I based this on the earlier flatbuffers of f142

amittal94 replied:
as you mentioned, this is required.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    ControlT.AddLimitHigh(builder, control_data.limitHigh)
+    ControlT.AddMinStep(builder, control_data.minStep)
+    control_offset = ControlT.End(builder)
+    return control_offset
```
```suggestion
    return ControlT.End(builder)
```

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+                builder, data
+            )
+        else:
+            data_start_offset, fb_type, enum_type = _serialise_numeric_scalar(
```
Somewhere around here we need to catch non-numeric dtypes (I think there's an `object` dtype). I guess you'll get a `KeyError`, but we want to give a clear `TypeError` to our users.

Same for arrays below.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    Returns:
+        The FlatBuffers offset for the serialised NTAttribute object.
+    """
+    if not ntattribute_data:
+        return 0
```
`safe_serialise` already does this.

amittal94 replied:
ah must have left it from a previous version

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    NTAttribute.Start(builder)
+    NTAttribute.AddName(builder, name_offset)
+    NTAttribute.AddValue(builder, value_offset)
+    NTAttribute.AddTags(builder, len(tags_offsets))
+    for tag_offset in reversed(tags_offsets):
+        NTAttribute.AddTags(builder, tag_offset)
```
I don't think this is right. Don't you need `StartTagsVector` and friends?

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+        codec=deserialise_codec(buffer.Codec()),
+        compressedSize=buffer.CompressedSize(),
+        uncompressedSize=buffer.UncompressedSize(),
+        dimension=[
+            dim
+            for i in range(buffer.DimensionLength())
+            if (dim := deserialise_dimension(buffer.Dimension(i))) is not None
+        ],
```
Hmm. Not sure about skipping the missing ones here. Attributes below are fine, but the position of a dimension within the list is very important.

amittal94 replied:
I kind of did this to make the typing work really, I see what you mean, but will we ever have such a case? Should we not expect dimensions to be correct or not provided at all?

amittal94 replied:
Ok so the only way we can support null values within our vector would be to use a union type and define some sort of null table. We will need to decide if this is worth it. Given we dont support serialising it, perhaps it may be best to raise an error in such a case. Also there should be validation checks pre-serialisation assuming we use the defined classes.

amittal94 replied:
I thought about it further, and this will line up with the base string arrays and other sorts as well, which would give an error otherwise

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    """
+    return dt.NTTable(
+        labels=[buffer.Labels(i).decode("utf-8") for i in range(buffer.LabelsLength())],
+        value=[
+            col
+            for i in range(buffer.ValueLength())
+            if (col := deserialise_column(buffer.Value(i))) is not None
+        ],
```
Again, we shouldn't skip missing values, because then they won't match up with the labels.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    Raises:
+        ValueError: If an unsupported or unknown data type is provided.
+    """
+    builder = flatbuffers.Builder(1024)
```
That's just an initial size I hope.

amittal94 replied:
yeah

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    elif isinstance(data.data, dt.NTTable):
+        data_offset = serialise_nttable(builder, data.data)
+        data_enum = PVType.PVType.NTTable
+    elif isinstance(data.data, dt.CAScalarAll):
+        converted_data = cascalarall_to_ntscalarall(data.data)
+        data_offset = serialise_ntscalarall(builder, converted_data)
+        data_enum = PVType.PVType.NTScalarAll
```
Oh I see! Not sure I'm a fan. Personally I think the conversion would be better if it was explicit.

amittal94 replied:
lets discuss this part.

amittal94 replied:
for the time being I have removed this from the flatbuffer serialisation/deserialisation

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+    PVData.AddPvName(builder, pv_name_offset)
+    pv_offset = PVData.End(builder)
+    builder.Finish(pv_offset, file_identifier=FILE_IDENTIFIER)
+    return bytes(builder.Output())
```
This may create a copy of the buffer

amittal94 replied:
which bit exactly, why would it create a copy?

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,1002 @@
+
+    if data_type == PVType.PVType.NTScalarAll:
+        ntscalarall_data = NTScalarAll.NTScalarAll()
+        ntscalarall_data.Init(data_buffer.Bytes, data_buffer.Pos)
```
I'm nearly at the end, and I have now seen all the horrors of the flatbuffers Python API that you have already experienced. This may be my favourite, for being a completely unnecessary usability hurdle.

stephen-jd-dann reviewed (commented):
I hope some of my comments are helpful. Although some of them are more venting about flatbuffers than actionable feedback.

stephen-jd-dann commented:
As the teeth marks in my desk will attest, I have just discovered that the `alarm.status` field in PVData does not have the same meaning as the status field in channel access. Here's some output from `pvxget`:

```
    value double = 9
    alarm.severity int32_t = 2
    alarm.status int32_t = 1
    alarm.message string = "HIHI"
```

The channel access status code for `HIHI` is 3, not 1. We see the same 1 status for `LOLO`, `LOW`, and `HIGH`. 1 here indicates `DEVICE`, as the output of `pvget` shows:

```
stephen:ai1 2025-02-13 15:42:09.129  9 MAJOR DEVICE HIHI
```

We'll need to apply a suitable conversion, and ideally fill in the alarm message at the same time. I refer you to the [relevant code in the PVXS library](https://github.com/epics-base/pvxs/blob/0ec2d3f6be5f00cb3b4f04d6adf7676570d6c79d/ioc/iocsource.cpp#L183-L238). This is the portion that runs inside an IOC and fills out a PVData Value object using the entries in the EPICS record database. As such, it uses some of the EPICS core libraries.

The status is handled with a hard-coded switch statement, but the message relies on an array of strings. We may have to find and copy it.

amittal94 commented:
I will make the suggested improvements best as I can. We can think of a better approach to handling channel access from what I gather. Switching to pydantic types was meant to be helpful, and standardise things, but we can also reconsider that if you think it does not work.

amittal94 committed:
Renaming and Refactorisation:

- rename AnyOuter to AnyT
- rename NTScalarAll to NTScalar
- sort imports
- simplify mapping dictionaries
- consistency improvements
- serialise empty array
- fix tags handling
- improve docstrings
- return more accurate error type
- remove cascalar handling from pva0_data
- rename cascalarall to cascalarany

- simplify serialise_data arguments

amittal94 commented:
I have done a few rebases, but history is mostly preserved

amittal94 committed:
removal of EnumT by using custom pydantic validation

amittal94 committed:
improved handling of none cases:

- any type handled with safe_serialise and safe_deserialise
- missing value field raises a ValueError
- cases of vectors with unusual values handled as errors

amittal94 committed:
initial testing for pva0

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,133 @@
+  description: string;
+  units: string;
+  precision: int;
+  form: string;
```
It would be nicer to make this a flatbuffers enum.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,131 @@
+    description: str = ""
+    units: str = ""
+    precision: int = 0
+    form: Annotated[str, BeforeValidator(extract_enum)] = ""
```
Per the suggestion in the fbs file, I think it would be nicer if this was an enum instead of a `str`. I know, technically you can put anything you like inside the PVData because the list of choices is defined at runtime, but the normative types spec is very clear about what the list of choices is supposed to be.

We'll need a little care with implementation. Enum variants in Python can be given values of any type, so we could just use the string values. Example:

```python
class FormT(Enum):
    DEFAULT = "Default"
    STRING = "String"
    ...
```

(I think the convention for enum variants is all-caps, but I'm not sure)

amittal94 replied:
Its not quite an python enum directly, but it is validated against the flatbuffers enum currently.

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -0,0 +1,979 @@
+        else:
+            data_start_offset, fb_type, enum_type = _serialise_array(builder, data)
+    else:
+        raise TypeError("Ony 0D and 1D Arrays are supported.")
```
Nitpicks:

```suggestion
        raise TypeError("only scalars and 1D Arrays are supported")
```

I think the usual convention for exception messages is that they start with a lowercase letter and don't end with a `.`

amittal94 replied:
ok will correct this and an others I find

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -0,0 +1,30 @@
+
+
+def cascalarany_to_ntscalarany(data: dt.CAScalarAny) -> dt.NTScalarAny:
+    alarm = dt.AlarmT(severity=data.severity, status=data.status)
```
Popping a comment in here to track the alarm status work, which we've already discussed. Maybe we need some sub-issues now?

amittal94 replied:
Can do so, but will still be on the same branch I guess.

amittal94 replied:
Although I guess I can create a branch from this branch meant to merged here as well actually

stephen-jd-dann replied:
Yes, I think this work is big enough that we should have been using an integration branch

stephen-jd-dann reviewed (commented):
Unless it's a lot of work, I'd like to implement form as an enum (a proper one, not that `enum_t` stuff) instead of a string.

I still need to go through unresolved conversations and add comments to them. Maybe we should go through them together tomorrow?

amittal94 commented:
> I still need to go through unresolved conversations and add comments to them. Maybe we should go through them together tomorrow?

Yeah sure.

amittal94 committed:
change form to a flatbuffer enum

amittal94 committed:
add AlarmStatus and AlarmSeverity Enums, and drop defaults

amittal94 committed:
updated custom validator for all enums

amittal94 committed:
ushort to ubyte for enums, and a TODO for optimisation

amittal94 committed:
general PyEpics converter

amittal94 committed:
Refactorisation and Reviewed Changes

- rename PyEpicsConverter to CatoNtConverter
- change CAAlarmStatus to CaAlarmStatus
- change the enum to an intenum
- handle the alarm status using the enum (required for proper typing)
- explicit message for an undefined alarm status
- data -> data.value (bug)
- rename the conversion function to convert_scalar
- modify the conversion function to also accept dictionaries
- rename status to ca_status where appropriate
- access AlarmStatus via datatypes to keep abstraction

amittal94 committed:
Various Improvements

- modify the convert scalar method to optionally ignore fields

- set defaults for cascalar

- make ca_to_pva kws keyword only

amittal94 committed:
unit tests for ca to pva

amittal94 committed:
follow better conventions and improve docstrings, fix missing descriptor

amittal94 commented:
This should be good to go mostly, bar some renaming for Acronym conventions, which I can do here, or in a seperate PR, or leave for now. I will likely improve upon this later too, maybe as part of the PR to add functionality to choose what to serialise.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,162 @@
+}
+
+table AlarmT {
+  severity: AlarmStatus;
+  status: AlarmSeverity;
```
Hang on a minute: these are backwards! How did neither of us notice that before?

amittal94 replied:
I flummoxed that this didnt result in any errors and wasn't caught before

stephen-jd-dann replied:
I had a look at the generated code in the `AlarmT.py` file. It never seems to check against the allowed enum values.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,162 @@
+
+table CodecT {
+  name: string;
+  parameters: AnyT;
```
Have you seen any examples of what this looks like in reality? I'm a little concerned about forward compatibility.

amittal94 replied:
only case I saw was name "" and parameters 5. I was unable to get any further input.

stephen-jd-dann replied:
I'm considering just leaving out the parameters field for now, until we've tried with a few real codecs. We can always add it back in later.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,162 @@
+
+table PVData {
+  data: PVType (required);
+  pv_name: string;
```
I've been wondering about renaming this to `source_name`, in case it's not a PV in future.

amittal94 replied:
can do so, what other sources could occur?

stephen-jd-dann replied:
We might have automated analysis programs running and producing data into Kafka. No actual PV name in that case. This sort of stream processing is a big selling point of Kafka.

stephen-jd-dann commented on schemas/pva0.fbs:
```diff
@@ -0,0 +1,162 @@
+  pv_name: string;
+}
+
+root_type PVData;
```
Nitpick: no newline at end of file

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,159 @@
+class AlarmT(BaseModel):
+    severity: Annotated[
+        int,
+        BeforeValidator(
```
It might be nice to have a shorthand for this, like `replace_none` above.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,159 @@
+
+
+class NTNDArray(BaseModel):
+    value: Any
```
It's not really `Any` though. It would be nice to have something more precise for all these, but I think that's a compatible change so it can wait until another PR.

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -0,0 +1,159 @@
+
+
+class PVData(BaseModel):
+    data: Union[NTScalarAny, NTNDArray, NTTable, CAScalarAny]
```
Can we remove `CAScalarAny` from this union? I don't think it's used any more.

stephen-jd-dann commented on tests/test_pva0.py:
```diff
@@ -0,0 +1,211 @@
+            ["test1", "test2", "test3"],  # string array
+            1,  # int
+            [1, 2, 3],  # int array
+            0x7F,  # byte
+            [0x7E, 0x7F, 0xFF],  # byte array
```
These are just ints, like the two cases above.

On reflection, I don't think you have any way of controlling the integer size used during serialisation. I'm not sure you can tell what size a scalar integer was when you get it from p4p. We might need to make a note of this as something to investigate later.

amittal94 replied:
would this be useful to know? for serialising to the smallest size possible? This may be another thing that we need benchmarking for as well

stephen-jd-dann replied:
Yes, we should probably test serialising to bytes specifically. As far as I can tell, the current interface doesn't let us specify what type we're serialising to.

stephen-jd-dann reviewed (changes requested):
Nearly there. A couple of minor issues, basically in making sure we've got the schema file right and won't need to make incompatible changes to it.

amittal94 committed:
Requested Changes and Improvements

- fixed severity and status swap in alarmt in pva0.fbs

- changed pv_name to source_name

- improved typing in pva0_data

- removed unnecessary tests

- removed CAScalarAny from PVData data union

- added tests for serialisation into bytes

- parse_enum wrapper for extract enum validator for better readability

- removed parameters from codec for the time being

stephen-jd-dann commented on src/epac/flatbuffers/data_types.py:
```diff
@@ -37,18 +37,19 @@ def extract_enum(enum_value: Optional[Any], enum_name: str, enum_type) -> int:
     return enum_value
 
 
+def validate_enum(enum_name: str, enum_type):
```
I'd nitpick the term "validate" here. This is really parsing, because you return a different value to the one passed in.

amittal94 replied:
Yeah I did not like the name either, Ill give it a rename

stephen-jd-dann commented on src/epac/flatbuffers/pva0_data.py:
```diff
@@ -271,7 +280,7 @@ def _serialise_array(builder: flatbuffers.Builder, data: np.ndarray):
 
 @safe_serialise
-def serialise_any(builder: flatbuffers.Builder, data) -> int:
+def serialise_any(builder: flatbuffers.Builder, data: Any) -> int:
```
We can make this type (`Any`) more specific later.

stephen-jd-dann reviewed (dismissed):
LGTM

amittal94 commented:
I will clean up the commits with a rebase, and merge it after approval

amittal94 commented:
renamed validate_enum to parse_enum, and no other changes besides some squashing some commits. Please let me know if you want to combine some more commits for clarity, or merge as is.

stephen-jd-dann reviewed (approved):
I think it's fine as is.

I normally suggest that you don't use commits like "requested changes" or "bug fixes". It's better to amend the specific commit that is being changed/fixed. However, that could be tricky here, and the commit history looks useful enough that I would say don't squash it.
